src/stats/dendrogram.jl

```diff
@@ -0,0 +1,123 @@
+"""
+    dendrogram(x, y; kwargs...)
+
+Draw a [dendrogram](https://en.wikipedia.org/wiki/Dendrogram),
+with leaf nodes specified by `x` and `y` coordinates,
+and parent nodes identified by `merges`.
+
+# Arguments
+- `x`: x positions of leaf nodes
+- `y`: y positions of leaf nodes (default = 0)
+# Keywords
+- `merges`: specifies connections between nodes (see below)
+- `treestyle`: one of `:??`, `:??` 
+"""
+@recipe(Dendrogram, x, y) do scene
+    Theme(
+        weights = automatic,
+        color = theme(scene, :patchcolor),
+        colormap = theme(scene, :colormap),
+        colorrange = automatic,
+        orientation = :vertical,
+        strokecolor = theme(scene, :patchstrokecolor),
+        strokewidth = theme(scene, :patchstrokewidth),
+        cycle = [:color => :patchcolor],
+        inspectable = theme(scene, :inspectable)
+    )
+end
+
+conversion_trait(x::Type{<:Dendrogram}) = SampleBased() #??
+
+function Makie.plot!(plot::Dendrogram)
+    args = @extract plot (weights, width, range, show_outliers, whiskerwidth, show_notch, orientation, gap, dodge, n_dodge, dodge_gap)
+
+end
+
+##############
+# Playground #
+##############
+
+struct DNode
+    idx::Int
+    x::Float32
+    y::Float32
+    children::Union{Tuple{Int,Int}, Nothing}
+end
+
+function find_merge(n1, n2; height=1)
+    newx = min(n1[1], n2[1]) + abs((n1[1] - n2[1])) / 2
+    newy = max(n1[2], n2[2]) + height
+    return Point2f(newx, newy)
+end
+
+function find_merge(n1::DNode, n2::DNode; height=1, index=max(n1.idx, n2.idx)+1)
+    newx = min(n1.x, n2.x) + abs((n1.x - n2.x)) / 2
+    newy = max(n1.y, n2.y) + height
+
+    return DNode(index, Point2f(newx, newy)..., (n1.idx, n2.idx))
+end
+
+
+function get_tree_connectors(parent, child1, child2)
+    return [(child1.x, child1.y), (parent.x, parent.y), (child2.x, child2.y)]
+end
+
+function get_box_connectors(parent, child1, child2)
+    yp = parent.y
+    x1 = child1.x
+    x2 = child2.x
+
+    return [(x1, child1.y), (x1, yp), (x2, yp), (x2, child2.y)]
+end
+
+function recursive_draw_dendrogram!(ax, node, nodes; branch_shape=:tree)
+    isnothing(node.children) && return nothing
+    child1 = nodes[node.children[1]]
+    child2 = nodes[node.children[2]]
+   
+    l = branch_shape == :tree ? get_tree_connectors(node, child1, child2) :
+        branch_shape == :box  ? get_box_connectors(node, child1, child2) : error()
+    
+    lines!(ax, l)
```

Comment:
Since Makie works on observables, every recipe has to be written with the idea that the underlying data may be updated.  `recursive_draw_dendrogram!` should return an array of points, and potentially an array of colors, then those would be plotted in a single `lines!` call in the recipe.